File: src/vos/tests/vos_size.py
#!/usr/bin/env python
'''
  (C) Copyright 2019 Intel Corporation.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
  The Government's rights to use, modify, reproduce, release, perform, display,
  or disclose this software are subject to the terms of the Apache License as
  provided in Contract No. B609815.
  Any reproduction of computer software, computer software documentation, or
  portions thereof marked with this legend must also reproduce the markings.

Prerequisite
Run vos_size to generate vos_size.yaml
Usage: vos_size.py input.yaml [alternate vos_size.yaml]

Sample input yaml
---
# yaml sample
num_pools: 10
updates:
  - array:
      rec_per_akey: 100
      avgsize: 1000000
      container: a
      object: a
      dkey: 20
      dkey_type: "integer_dkey"
      akey_per_dkey: 1
  - single_value:
      avgsize: 512000
      container: a
      object: a
      dkey: 20
      akey_per_dkey: 10
'''
import sys
import yaml

def print_total(name, pool_stats):
    "Pretty print"
    stat = pool_stats[name]
    total = pool_stats["total"]

    print "\t%-20s: %10d K (%5.2f%%)" % (name, stat / 1024,
                                         100 * float(stat) / total)

# pylint: disable=too-many-instance-attributes
class MetaOverhead(object):
    """Class for calculating overheads"""
    def __init__(self, num_pools, meta_yaml):
        """class for keeping track of overheads"""
        self.meta = meta_yaml
        self.num_pools = num_pools
        self.pools = []
        for _index in range(0, self.num_pools):
            self.pools.append({"key" : "container", "count" : 0})
        self.next_cont = 1341344
        self.next_object = 1341344
        self.value_size = 0
        self.unique_entries = 0
        self.container_key = None

    def init_container(self, array_spec):
        """ Ensure the existence of container records"""
        if "container" in array_spec:
            key = array_spec["container"]
        else:
            key = self.next_cont
            self.next_cont += 1

        for pool in self.pools:
            if key not in pool:
                pool[key] = {"key" : "object", "count": 0} #object count
                pool["count"] += 1

        self.container_key = key

    def update_object(self, array_spec, dkey_type, pool_num):
        """Ensure existence of the object"""
        cont = self.pools[pool_num][self.container_key]
        if "object" in array_spec:
            key = array_spec["object"]
        else:
            key = self.next_object
            self.next_object += 1

        if key not in cont:
            cont[key] = {"key" : dkey_type, "count": 0} #dkey count
            cont["count"] += 1

        return cont[key]

    def update_value(self, array_spec, value_type):
        """Update the value according to the spec"""
        dkey_type = array_spec.get("dkey_type", "dkey")
        akey_type = array_spec.get("akey_type", "akey")

        dkeys = int(array_spec.get("dkey", 1))
        akeys = int(array_spec.get("akey_per_dkey", 1))
        records = int(array_spec.get("rec_per_akey", 1))

        for dkey in range(0, dkeys):
            pool_num = (dkey % self.num_pools)
            obj = self.update_object(array_spec, dkey_type, pool_num)
            if dkey not in obj:
                obj[dkey] = {"key": akey_type, "count": 0} #akey count
                obj["count"] += 1
            dkey_tree = obj[dkey]
            for akey in range(0, akeys):
                if akey not in dkey_tree:
                    dkey_tree[akey] = {"key" : value_type, "count": 0}
                    dkey_tree["count"] += 1
                akey_tree = dkey_tree[akey]
                akey_tree["count"] += records
        unique = dkeys * akeys * records
        self.unique_entries += unique
        self.value_size += unique * int(array_spec.get("avgsize"))

    def update(self, value_type, array_spec):
        """calculate metadata for update(s)"""
        self.init_container(array_spec)
        self.update_value(array_spec, value_type)

    def calc_tree(self, tree, pool_stats):
        """calculate the totals"""
        key = tree["key"]
        num_values = tree["count"]
        record_size = self.meta["trees"][key]["record_msize"]
        node_size = self.meta["trees"][key]["node_size"]
        order = self.meta["trees"][key]["order"]
        rec_overhead = num_values * record_size
        tree_nodes = (num_values * 2 + order -1) / order
        overhead = tree_nodes * node_size + rec_overhead
        key = key.replace("integer_", "")
        pool_stats[key] += overhead
        pool_stats["total"] += overhead
        pool_stats["total_meta"] += overhead
        if key == "array" or key == "single_value":
            return
        for key in tree.keys():
            if key == "count":
                continue
            if key == "key":
                continue
            self.calc_tree(tree[key], pool_stats)

    def print_report(self):
        """Calculate and pretty print a report"""
        pool_stats = {"pool" : 0,
                      "container" : 0,
                      "object" : 0,
                      "dkey" : 0,
                      "akey" : 0,
                      "array" : 0,
                      "single_value" : 0,
                      "total_meta" : 0,
                      "total" : self.value_size
                     }

        for pool in range(0, self.num_pools):
            pool_stats["pool"] += int(self.meta.get("root"))
            pool_stats["total_meta"] += int(self.meta.get("root"))
            pool_stats["total"] += int(self.meta.get("root"))
            self.calc_tree(self.pools[pool], pool_stats)

        print "Metadata totals:"
        print_total("pool", pool_stats)
        print_total("container", pool_stats)
        print_total("object", pool_stats)
        print_total("dkey", pool_stats)
        print_total("akey", pool_stats)
        print_total("single_value", pool_stats)
        print_total("array", pool_stats)
        print_total("total_meta", pool_stats)
        print "Total bytes with user data: %dK" % (pool_stats["total"] / 1024)

# pylint: enable=too-many-instance-attributes

def run_vos_size():
    """Run the tool"""

    if len(sys.argv) < 2:
        print "Usage: %s <configuration> [<meta config>]" % sys.argv[0]
        sys.exit(-1)

    config_name = open(sys.argv[1], "r")
    config_yaml = yaml.load(config_name)

    meta_name = "vos_size.yaml"
    if len(sys.argv) == 3:
        meta_name = sys.argv[2]
    meta_yaml = yaml.load(open(meta_name, "r"))

    num_pools = config_yaml.get("num_pools", 1)

    overheads = MetaOverhead(num_pools, meta_yaml)

    for update in config_yaml.get("updates"):
        for update_type in update.keys():
            if update_type == "array" or update_type == "single_value":
                overheads.update(update_type, update[update_type])

    overheads.print_report()

if __name__ == "__main__":
    run_vos_size()

File: src/vos/tests/vos_size.c
/**
 * (C) Copyright 2019-2020 Intel Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
 * The Government's rights to use, modify, reproduce, release, perform, display,
 * or disclose this software are subject to the terms of the Apache License as
 * provided in Contract No. B609815.
 * Any reproduction of computer software, computer software documentation, or
 * portions thereof marked with this legend must also reproduce the markings.
 */
/* This generates a summary of struct sizes to be used by vos_size.py
 * to generate metadata overhead estimates
 */
#define D_LOGFAC	DD_FAC(tests)

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include <daos/debug.h>
#include <daos_srv/vos.h>

static int alloc_overhead = 16;

#define FOREACH_TYPE(ACTION)						\
	ACTION(container, VOS_TC_CONTAINER, 0)				\
	ACTION(object, VOS_TC_OBJECT, 0)				\
	ACTION(dkey, VOS_TC_DKEY, 0)					\
	ACTION(akey, VOS_TC_AKEY, 0)					\
	ACTION(integer_dkey, VOS_TC_DKEY, BTR_FEAT_DIRECT_KEY)		\
	ACTION(integer_akey, VOS_TC_AKEY, BTR_FEAT_DIRECT_KEY)		\
	ACTION(single_value, VOS_TC_SV, 0)				\
	ACTION(array, VOS_TC_ARRAY, 0)

#define DECLARE_TYPE(name, type, feats)	\
	struct daos_tree_overhead	name;

#define CHECK_CALL(name, type, feats)					\
	do {								\
		rc = vos_tree_get_overhead(alloc_overhead, type, feats,	\
					   &name);			\
		if (rc != 0) {						\
			printf(#name " lookup failed: rc = "DF_RC"\n",	\
				DP_RC(rc));				\
			goto exit_1;					\
		}							\
	} while (0);

#define PRINT_DYNAMIC(name, type, feats)				\
	print_dynamic(fp, #name, &name);

#define PRINT_RECORD(name, type, feats)					\
	print_record(fp, #name, &name);

static void
print_dynamic(FILE *fp, const char *name, const struct daos_tree_overhead *ovhd)
{
	int	i;

	if (ovhd->to_dyn_count == 0)
		return;

	for (i = 0; i < ovhd->to_dyn_count; i++) {
		fprintf(fp, "%s_%d_key: &%s_%d\n", name,
			ovhd->to_dyn_overhead[i].no_order, name,
			ovhd->to_dyn_overhead[i].no_order);
		fprintf(fp, "  order: %d\n", ovhd->to_dyn_overhead[i].no_order);
		fprintf(fp, "  size: %d\n", ovhd->to_dyn_overhead[i].no_size);
	}
}

static void
print_record(FILE *fp, const char *name, const struct daos_tree_overhead *ovhd)
{
	int	i;
	int	count = 0;

	fprintf(fp, "  %s:\n", name);
	fprintf(fp, "    order: %d\n", ovhd->to_leaf_overhead.no_order);
	fprintf(fp, "    leaf_node_size: %d\n", ovhd->to_leaf_overhead.no_size);
	fprintf(fp, "    int_node_size: %d\n", ovhd->to_int_node_size);
	fprintf(fp, "    record_msize: %d\n", ovhd->to_record_msize);
	fprintf(fp, "    node_rec_msize: %d\n", ovhd->to_node_rec_msize);
	fprintf(fp, "    num_dynamic: %d\n", ovhd->to_dyn_count);
	if (ovhd->to_dyn_count == 0)
		return;

	fprintf(fp, "    dynamic: [\n      ");
	for (i = 0; i < ovhd->to_dyn_count; i++) {
		count += fprintf(fp, "*%s_%d", name,
				 ovhd->to_dyn_overhead[i].no_order);
		if (i == (ovhd->to_dyn_count - 1))
			continue;
		if (count > 40) {
			count = 0;
			fprintf(fp, ",\n      ");
		} else {
			count += fprintf(fp, ", ");
		}
	}
	fprintf(fp, "\n    ]\n");
}


char *
alloc_fname(const char *requested)
{
	char *suffix;
	char *fname;

	suffix = strstr(requested, ".yaml");

	if (!suffix || strcmp(suffix, ".yaml") != 0)
		D_ASPRINTF(fname, "%s.yaml", requested);
	else
		D_STRNDUP(fname, requested, PATH_MAX);

	if (fname == NULL)
		printf("Could not allocate memory to save %s\n", requested);

	return fname;
}

void
free_fname(char *fname)
{
	D_FREE(fname);
}

FILE *
open_file(const char *fname)
{
	FILE		*fp;
	const char	*to_open = fname;

	if (fname == NULL)
		to_open = "vos_size.yaml";

	fp = fopen(to_open, "w");
	if (fp == NULL)
		printf("Could not open %s: %s\n", to_open, strerror(errno));

	return fp;
}

void
print_usage(const char *name)
{
	printf("Usage: %s [OPTIONS]\n\
	OPTIONS:\n\
		--alloc_overhead, -a <bytes>	Overhead bytes per alloc (16)\n\
		--fname, -f <filename>		Output file (vos_size.yaml)\n\
		-h				Print this help message\n",
		name);
}

int
main(int argc, char **argv)
{
	FILE				*fp;
	char				*fname = NULL;
	static struct option		 long_options[] = {
		{"fname",		required_argument, 0, 'f'},
		{"help",		no_argument, 0, 'h'},
		{"alloc_overhead",	required_argument, 0, 'a'},
	};
	FOREACH_TYPE(DECLARE_TYPE)
	int				 rc;
	int				 index = 0;
	int				 opt = 0;

	rc = daos_debug_init(DAOS_LOG_DEFAULT);
	if (rc) {
		printf("Error initializing debug system\n");
		return rc;
	}

	rc = vos_init();
	if (rc) {
		printf("Error initializing VOS instance\n");
		goto exit_0;
	}

	while ((opt = getopt_long(argc, argv, "a:f:h",
				  long_options, &index)) != -1) {
		switch (opt) {
		case 'a':
			alloc_overhead = atoi(optarg);
			break;
		case 'f':
			fname = alloc_fname(optarg);
			if (fname == NULL)
				goto exit_1;
			break;
		case 'h':
			print_usage(argv[0]);
			goto exit_1;
		default:
			printf("Unknown option\n");
			print_usage(argv[0]);
			goto exit_1;
		}
	}

	FOREACH_TYPE(CHECK_CALL)

	fp = open_file(fname);
	free_fname(fname);

	if (fp == NULL)
		goto exit_1;

	fprintf(fp, "---\n# VOS tree overheads\n"
		"root: %d\nscm_cutoff: %d\n", vos_pool_get_msize(),
		vos_pool_get_scm_cutoff());

	FOREACH_TYPE(PRINT_DYNAMIC)
	fprintf(fp, "trees:\n");
	FOREACH_TYPE(PRINT_RECORD)

	fclose(fp);
exit_1:
	vos_fini();
exit_0:
	daos_debug_fini();
	return 0;
}


